### DISCUSSION TITLE
1.0 Roadmap

### DISCUSSION THREAD

**Discussion by jonatkinson** [dokku/dokku#4258] [General]
Hello.

I'm a user of Dokku for my personal projects (for a longer than I care to remember!), and I'm investigating the possibility of introducing Dokku at my company. Part of this decision involves some understanding of the future plans for Dokku. I understand there is currently a `1.0` milestone (which is useful), and also there seems to be a longer-term goal of rewriting in Golang (which I've picked up from various scattered references).

I was wondering if there was a more general roadmap document, which outlines in more detail the plans for 1.0 (and beyond). Maybe there is already but it's not clearly signposted? Any information would be welcome.

**Comment by josegonzalez** [comment]
Bit of a rant, but hopefully the following helps you make a decision as to whether Dokku is good enough for your organization.

---

There isn't really a roadmap, mostly because the end state has shifted over time. Generally, the project is trending towards the following:

- Heroku Compatibility
- The best single-server platform for code deployments
- Flexibility around defaults and the ability to add/replace core components

### Goals

The stated goal of "General Heroku Compatibility" is still there. That said:

- We do support functionality Heroku does not and never will (alternative schedulers, custom plugins)
- Some things are hard to support since we've mostly eschewed having daemons (other than the [dokku-event-listener](https://github.com/dokku/dokku-event-listener)), such as http(s) based deployment or an api.
- Some functionality will never be supported because they are a pain in the ass and better suited to external projects. Heroku datastore addons are Dokku plugins, and we don't have a fancy routing layer, just plain old Nginx.

We also generally want to be the best single-server platform for deploying code. We're not necessarily the simplest - `git pull` + some random bash commands might be simpler for your use case - I do want it to be the most hands-off in the general case. Having someone install Dokku with a few commands (curl bash or otherwise) and then being able to `git push dokku master` to get their app deployed in under 5 minutes is a great experience.

I'd also like it continue to be "batteries included, but replaceable". The Dokku defaults are great, but aren't always going to be what people want. Heroku buildpacks are not a panacea, so you also get Dockerfile and Docker Image support. Our plugins won't always support every feature you want, but you should be able to build custom ones (or even replace the core ones!). This comes at some complexity (both in terms of Dokku core code as well as your own setup), but they are options for cases where Dokku is 80-90% of the solution, but you need just a tiny bit more to get everything setup nicely.

### Is using an OSS project good for you?

Dokku isn't backed by any financial organization - it's an OSS project that I work on in my spare time. I don't expect this to change much (though see the next section for some thoughts here). Others have come and gone - and I welcome new contributors! - but I think it takes a special kind of crazy to work on deployment tools. There are other tools and services out there that you could use instead of Dokku, and I don't pretend to think that Dokku will be the best solution even in the majority of cases that we're targeting.

Maybe using Heroku directly would be safer for you/your org, or maybe you should build a more focused deployment tool internally on top of some building blocks from your hosting provider. Maybe your hosting provider has a good solution you can use directly? There is always risk in using an OSS project to run a business, and it's worth thinking about whether the risk is worth it. Maybe I get bored in the future and stop working on Dokku, or I start working for an organization that refuses to allow me to work on Dokku or certain functionality as part of the OSS offering.

But honestly, you run the same risk by using something that _is_ serviced by a company or has some financial backing - feel free to do a little research on the number of PaaS companies out there that have been acquired and shut down or have just gone out of business. I did some of this research a few years back when considering trying to build a startup behind Dokku, and figured it was just too risky to justify attempting to do so.

### Is Dokku stable for production use?

Yes, and there are quite a few organizations either using it in production or using it as part of a product. I won't comment on who, but they are out there in surprising numbers.

That said, Dokku does indeed bugs. What software does not? It would be great to not have any bugs. If a bug hits your usage, I promise to try and fix it, but it's OSS, so that'll be limited by:

- whether I can easily reproduce your bug (a bats test case is great for this!)
- if I have time to fix it in between work and personal stuff
- if I am interested in fixing it, because not all problems are interesting and sometimes a hard problem is not something I can knock out before dinner
- if you can [grease the wheels with money](https://github.com/dokku/.github/blob/master/SPONSORING.md)

### Can I make money from this thing?

I intend to start _selling_ an Api/Git Server/Manager binary in the near future. [Here is an early gif](https://twitter.com/dokku/status/1334801381778219008) of it in action. My plan is to sell it as part of a yearly subscription where folks get access to all updates during their subscription lifetime, with maybe volume licensing if you want to manage more than one server/cluster with separate installations. I would add functionality once a quarter or so, slowly exposing more uncommon functionality, but also making certain operations easier.

The idea here is to allow me to justify spending my free time on Dokku. The OpenCollective and Patreon have been ~fine but haven't really pushed the needle for me, and are certainly do not support the amount of time I've spent on Dokku over the years. There's a ton of issues here and for official plugins that I just haven't had time to investigate, and this would be a way of subsidizing it. One nice thing is that functionality in the package has required functionality to be developed upstream in Dokku itself, so I think it's overall a good thing for the project's health.

I've attempted similar in the past by selling specific plugins - the registry plugin was originally closed source, as currently are a git-sync and consul plugin - and the ansible modules. Neither has stuck, and they've been open sourced as a result. That said, if this doesn't work, I'll just stop working on it and keep it private. It won't stop me from working on Dokku - I'll still have need of the project unless Heroku open sources their platform and makes it a single binary - but will be a good signal for myself in terms of spending free time on the project.

### What is in a 1.0?

Aside from the above, I'd love to eventually get to a 1.0 with no - fixable - reported bugs. I'll have removed all the deprecated code. I think most/all of the plugins will be rewritten in Golang at that point, and the end-user interface will be relatively stable. Once we hit a 1.0, I'll mostly be focusing on bug fixes and test additions rather than new functionality.

Thats not to say there won't be any new features, but that most of the core workflow functionality will be set in stone and folks won't really need to worry so much about a new Dokku release breaking something about their deployment pipeline.

### Is there going to be a 2.0?

There will probably be a 2.0. It'll probably focus on internal refactoring to make it easier to use Dokku as a library and tighten up tooling usage.  I'd love to add support for ECS, Lambda, Cloud Run, etc. I think that most deployment tooling is targeted towards operators and not end-users, and Dokku is in a good position to be the porcelain over that plumbing.

It'll also be a good time to step back and really clean up the official plugins. Maybe having official plugins for RDS or similar will be in the cards. Not really sure, but I figure I'll have enough breathing room to tackle this.

### Can I have a roadmap?

If I can get Dokku funded, yes! But as it stands, I have general "plans" for the OSS project, and those plans will definitely change over time, so I don't think a roadmap makes so much sense.

**Reply by jonatkinson** [reply]
@josegonzalez Thank you. Both for your work on Dokku, and for your considered reply here. 

As an organisation, we are at peace with depending on OSS, and the inherent instability that implies (though, as you identified, we have had dependencies on commercial products in the past, some of which have been sunsetted with various levels of disruption). 

I also applaud your attempts to make this project sustainable; I'm sorry none have quite stuck the landing yet. For what it's worth, some web-based UI over Dokku would be helpful in my organisation; we have software engineers of various preferences, and not all are *nix literate enough to want to manage everything via CLI tools.

So thank you again, even as a rant this was very helpful.

**Reply by kidGodzilla** [reply]
Thanks for sharing this, @josegonzalez – I've been a big fan of Dokku for quite a while now. FWIW I think it reached a 1.0 level of stability and feature completeness for the average use-case a long while ago.

**Reply by badsyntax** [reply]
@josegonzalez what's the reason for moving to golang? I don't dissagee with this decision, but I loved that dokku was mostly bash as it was easy to read. Go is easy to read, but not as easy as bash imo. Either way, just wondering about your reason for the move.

**Reply by josegonzalez** [reply]
There are more than a few things that are difficult to do in Bash - bordering on what feels like impossible sometimes - that are super straightforward in Golang:

- Argument parsing
- JSON handling (reading and writing)
- Network requests
- Interacting with APIs
- Managing structured data

While all of these have been performed one time or another in bash, the fact is that there are a ton of edge cases that have popped up over the years that make it super difficult to even reason about some of the code, whereas it is fairly easy for me to handle these within Golang (either with the standard library or something off-the-shelf).

One other thing that was brought up early on was the idea that testing internal golang functions was easier than doing so for shell code. I haven't really gone down the route of unit testing in favor of continuing our integration tests. I hope to have some time do that more this year.

For what its worth, I don't think _every_ plugin needs to be rewritten, just the ones that stretch what we can do easily in Shell. The `git` plugin is a good example of something that will likely stay in Git, whereas new plugins will likely continue to be written in Golang. Yeah, I could probably write some helper binaries to make them more easily consumable in shell code, but my attempts at that haven't been too awesome thus far :(

**Reply by badsyntax** [reply]
Thanks for the detailed response @josegonzalez! I've experienced a bit of this pain in one of my plugins, and found it easier & cleaner to use python to perform a "complex" task (updating structured data in yaml format) than piping data into various cmd line tools. If I do another complex plugin I'll give Golang a bash.

**Comment by hiepxanh** [comment]
I think the roadmap people think is different, it's not your responsibility to finish these feature on that 1.0 version. If you think your OSS like a big project. And all these people over out there could make some help to Dokku. They need some task, some guideline, some kind of list think need to do, so they will know that they can make some help, for example:

Create a pin repository name: Dokku 1.0 plan (Yes a plan, a plan for community, and it will move as you like)

- What is coming ?
  + feature A
  + feature B
  + feature C
all this feature is what you think it should be in 1.0, not what it must be, it your project so you can drop it if you like as dokku destroy 😄 


- What are you doing ?

  + refactoring all kind of things 
  + adding plugin XXX (WIP)
  + adding multi Dockerfile deploy for monorepo (WIP but need help)
  + stuck at feature C (issues #xxx) 
  

- What is not coming ?
   + adding K8 (nice to have, PR welcome but It wont be there at this time, saving your time to ask) 
  

- What are you planing to do?
   + adding metric system (just plan, not doing anything, PR welcome, more discusss on #xxx I will give any api you need to write that feature)
   + writing docs for deploy static website (It necessary but I dont like to write it, PR is nice)
   + making UI interface for Dokku ( writen by React, Yahoo, everybody like this feature, more disscuss on #xyz)
 

 - need help?
   + writing docs on ABC
   + writing docs on XYZ
 
 That all. 
 
 Wait, dont feel stress on this,  as a newbie, I would like to help Dokku, If my bussiness get money, I will donate for team, if I dont have money, I think that kind of post can give me good report about current state of project, which one I can help, and which one I cant.
 
 I can make some kind of help like writing docs, but what kind of docs is missing? Or I can help you writing dokku UI interface with you, but what kind of Frontend Framework are you using? 
 
 Dokku is a great project, and a second pining issues list task should be done is neccessary to give community a guiden, a purpose, for who can help, can contribute it. What are you need to help you on that feature? 
 
 p/s: I think github sponsor is nice, you should add it too. [read this post](https://aeplay.org/citybound-devblog/why-im-moving-from-patreon-to-github-sponsors#:~:text=At%20the%20same%20time%2C%20the,for%20creators%20are%20really%20lacklustre.&text=GitHub%20Sponsors%20is%20like%20a,that%20already%20are%20on%20GitHub.)
 
 p/s 2: the first think I will do when my Business start making money is get a party, and donate for [Postgraphile](https://github.com/graphile/postgraphile) and This project.

**Reply by hiepxanh** [reply]
![image](https://user-images.githubusercontent.com/16970990/103970059-ec39f700-5199-11eb-9035-e1959f20388a.png)
LOL really ? you should try this man

**Reply by josegonzalez** [reply]
Thats all well and good, but project management on that level takes a lot of effort. I already put quite a serious amount of time into Dokku with very little recompense. The overhead of building a roadmap that would change over time isn't worth it if I can spend that time writing code that solves actual user problems. I say this having managed those types of processes over the years at various companies.

The [release board](https://github.com/dokku/dokku/projects/1) should give people a good idea as to work I'm working on in conjunction with the [1.0.0 milesetone](https://github.com/dokku/dokku/milestone/16).

---

Regarding folks contributing:

- There is a [contributing](https://github.com/dokku/dokku/blob/master/CONTRIBUTING.md) doc that outlines a ton of ways to contribute, not all with code or money.
- I try to triage every issue on the core project as quickly as possible (though certainly some slip through the cracks). I've labeled what type of issue it is (docs, bug, enhancement), which have or need a plan, the potential difficulty level, and even sometimes the merge level. Not sure what else I can do to help folks who want to contribute... contribute.
- I don't want external help at this time for the UI project, but I thank you for your offer. There are other projects like this out there (feel free to keep using them), but this one is something I'm working on to actually help justify my time on the project. Having others who haven't been involved in Dokku get involved in that project feels kinda like a slap to the face (do I give it to that person for free? are they expecting it to be foss? etc.).

This might come off as defensive, but I've done quite a bit to try and get others involved in Dokku, though significant PRs are rare. I appreciate all changes (and try to merge them as they come if I can test them), but not sure what I can do to encourage folks to contribute.

----

Regarding github sponsors, I'll check it out tomorrow. I didn't realize you could use existing sponsorships + github sponsors. I guess we'll see what happens, thanks for the info.

**Reply by hiepxanh** [reply]
For the time managment, and project, and family. I think you should read this post of Benji [README PROJECT]
(https://github.com/readme/benjie-gillam) I think he just normal man, and he have a great project. But donate him? maybe???

But after reading this post. I have an other view. I want to give him more money. I see how he manage to survive, how he can live with that little of money? what it can help for his wife? his kids? How he spend time for the OSS ? what did he wrong at the first time when he put too much effort and lossing balance? 
Dokku => help organize server
This post => I hope it can help you organize yourself, take balance 
If you have children, this post is basicly what are you feeling now, if you not. It is your feauture. His way is very useful.
whatever, I feel, me, myself have more responsibilty to countribute my money from the project which using your OSS. I feel that I can help a family, I think I can do something for you, more specifically about where my money is going to ? Who will using it? 
More than support a software, I'm feeling I'm support for a man, for a family, for a great Idea. And I want to give you my money more after reading that post instead of donate for few sexy baby girl on Patreon.
Writing and sharing "README PROJECT" is very useful. It help you reach your backer better,and it help me to understand the OSS leader better. 
You should write it and share this. After read that, maybe people want to donate you much more, be cause they understand you better

p/s: for an old rich Uncle, see that Patreon donate gateway? Maybe
but Github Sponsor? Yes, I will. Because I feel more professional, more safe payment gateway, feeling I'm not paying for adult site LOL

**Reply by hiepxanh** [reply]
@josegonzalez

**Reply by hiepxanh** [reply]
@josegonzalez

**Reply by josegonzalez** [reply]
We've finally been approved for Github Sponsors and a link is available [here](https://github.com/sponsors/dokku).

**Reply by hiepxanh** [reply]
![Screen Shot 2021-02-09 at 10 28 34](https://user-images.githubusercontent.com/16970990/107311870-c453f100-6ac1-11eb-862e-a14e0dd9a4c3.png)
not have much money, but I'm glad to sponsor you, cheer @josegonzalez

**Reply by hiepxanh** [reply]
@josegonzalez can you create more tier? 
50$ 100$ 150$ is very useful, I would donate more if i have more

**Reply by josegonzalez** [reply]
Wait, you can only donate at specific tiers and not a free-for-all? Interesting.

**Reply by josegonzalez** [reply]
I added two more tiers. Thanks for the advice :)

**Reply by hiepxanh** [reply]
you're welcome happy to donating for this project
